src/OrchardCore.Modules/OrchardCore.AutoSetup/AutoSetupMiddleware.cs

```diff
@@ -146,123 +138,28 @@ public async Task InvokeAsync(HttpContext httpContext)
                         {
                             if (_setupOptions != setupOptions)
                             {
-                                await CreateTenantSettingsAsync(setupOptions);
+                                await autoSetupService.CreateTenantSettingsAsync(setupOptions);
                             }
                         }
                     }
 
                     httpContext.Response.Redirect(pathBase);
+                }
+                else
+                {
+                    httpContext.Response.StatusCode = StatusCodes.Status503ServiceUnavailable;
+                    var stringBuilder = new StringBuilder();
+                    foreach (var error in setupContext.Errors)
+                    {
+                        stringBuilder.AppendLine($"{error.Key} : '{error.Value}'");
+                    }
 
+                    await httpContext.Response.WriteAsync($"The AutoSetup failed installing the site '{_setupOptions.SiteName}' with errors: {stringBuilder}.");
```

Comment:
You shouldn't write error messages to the response directly. For instance here the SiteName is user input and could contain XSS. I believe it was only logged before. If we want to keep doing it then we don't need the tuple as a return type.

Reply:
@sebastienros : Committed new one of shouldn't write error messages to the response directly